File: Tools/verifyBin.py
import os
import sys

# Define colors for output. This class borrowed from https://stackoverflow.com/a/287944/9710543
class bcolors:
    HEADER = "\033[95m"
    OKBLUE = "\033[94m"
    OKGREEN = "\033[92m"
    WARNING = "\033[93m"
    FAIL = "\033[91m"
    ENDC = "\033[0m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"

def get_files(dirName):
    dirList = os.listdir(dirName)
    files = []
    for entry in dirList:
        path = os.path.join(dirName, entry)
        if os.path.isdir(path):
            files = files + get_files(path)
        else:
            files.append(path)
                
    return files

if os.path.isdir(os.path.normpath("./out/Electron Style-win32-x64")):
    source = get_files(os.path.normpath("src/bin"))
    packaged = get_files(os.path.normpath("out/Electron Style-win32-x64/resources/app.asar.unpacked/src/bin"))

    for p in range(len(packaged)):
        packaged[p] = packaged[p][packaged[p].find("src"):]

    missing = []
    for i in range(len(source)):
        try:
            packaged.index(source[i])
        except ValueError:
            missing.append(source[i])

    if len(missing) != 0:
        message = "Missing: "
        for m in range(len(missing)):
            message = message + missing[m]
        print(bcolors.FAIL + str(len(packaged)) + "/" + str(len(source)) + " files packaged." + bcolors.ENDC)
        print(bcolors.FAIL + message + bcolors.ENDC + "\n")
        sys.exit(1)
    else:
        print(bcolors.OKGREEN + str(len(packaged)) + "/" + str(len(source)) + " files packaged." + bcolors.ENDC + "\n")
else:
    print(bcolors.FAIL + "No packaged directory found at: ./out/Electron Style-win32-x64" + bcolors.ENDC + "\n")
    sys.exit(1)

File: Tools/verifyBin.js
const fs = require('fs');
const path = require('path');

require('colors');

function walkSync(dir, filelist) {
    var fs = fs || require('fs'),
        files = fs.readdirSync(dir);
    filelist = filelist || [];
    files.forEach(function(file) {
      if (fs.statSync(dir + '/' + file).isDirectory()) {
        filelist = walkSync(dir + '/' + file + '/', filelist);
      }
      else {
        filelist.push(file);
      }
    });
    return filelist;
}

/**
 * @function compare_files
 * @description Compares the contents of two directories.
 * @param {*} dir1 String. Path to parent directory.
 * @param {*} dir2 String. Path to child directory.
 * @returns A list of missing files, if no files are missing the list will be empty.
 */
function compare_files(dir1, dir2) {
    const fs = require('fs');

    require('colors');

    const parent_files = walkSync(dir1);
    const child_files = walkSync(dir2);

    let missing = [];
    for (var i=0; i < parent_files.length; i++) {
        if (!child_files.includes(parent_files[i])) {
            missing.push(parent_files[i]);
        }
    }
    return missing;
}

if (fs.existsSync("./out/Electron Style-win32-x64")) {
    if (fs.existsSync("./src/bin/neural_style")) {
        var missing = compare_files("./src/bin", "./out/Electron Style-win32-x64");
        if (missing.length !== 0) {
            let fail_message = "Missing files: ";
            for (var i=0; i < missing.length; i++) {
                if (i < missing.length - 1) {
                    fail_message += missing[i] + " ";
                } else {
                    fail_message += missing[i];
                }
            }
            
            console.error(fail_message.red + "\n");
            process.exit(1);
        } else {
            console.log("All files packaged.".green + "\n");
            process.exit(0);
        }
    } else {
        console.error("No compiled neural style found at ./src/bin/neural_style".red + "\n");
        process.exit(1);
    }
} else {
    console.error("No packaged project found at ./out/Electron Style-win32-x64".red + "\n");
    process.exit(1);
}
